src/backend/optimizer/plan/createplan.c

```diff
@@ -781,13 +781,15 @@ create_join_plan(PlannerInfo *root, JoinPath *best_path)
 	Assert(plan->type == best_path->path.pathtype);
 	if (partition_selector_created)
 		((Join *) plan)->prefetch_inner = true;
-
-	/* CDB: if the join's locus is bottleneck which means the
-	 * join gang only contains one process, so there is no
-	 * risk for motion deadlock.
-	 */
-	if (CdbPathLocus_IsBottleneck(best_path->path.locus))
+	else if (CdbPathLocus_IsBottleneck(best_path->path.locus) &&
+		!(root->config->gp_cte_sharing && IsA(plan, HashJoin)))
```

Comment:
Well, if we want to preserve the behaviour when we disable the prefetch for Bottleneck in all cases but Shared Scans, we can't use that `root->config->gp_cte_sharing`. We may enable gp_cte_sharing but don't use CTE at all. And what, the prefetch will be enabled. And this will lead to performance regress as well. 
```
explain (costs off) select * from t_repl left join (select * from t_hashed limit 1) x using (i);
                            QUERY PLAN                            
------------------------------------------------------------------
 Hash Left Join
   Hash Cond: (t_repl.i = x.i)
   ->  Gather Motion 1:1  (slice1; segments: 1)
         ->  Seq Scan on t_repl
   ->  Hash
         ->  Subquery Scan on x
               ->  Limit
                     ->  Gather Motion 3:1  (slice2; segments: 3)
                           ->  Limit
                                 ->  Seq Scan on t_hashed
 Optimizer: Postgres query optimizer
(11 rows)

```

Reply:
Alternatively, we can make a walker to search for Shared Scans in the underlying nodes, or add a flag to PlannerInfo that will turn on when we create Shared Scans.

Reply:
Walker should be written only when we won't find other options. Investigate, please, the following proposal:
1. Here we allow dsabling the prefetch.
2. In apply_shareinput_dag_to_tree's shareinput_walker when we meet join, we save the prefetch state. Then for the [HashJoin](https://github.com/arenadata/gpdb/blob/d10d2bdea3f7a6df667d8fba60dd3d861f077883/src/backend/cdb/cdbmutate.c#L1909) we  enable the prefetch. 
3. If along the way we didn't meet the SharedScan, we restore the prefetch for the join with the saved value, otherwise the prefetch is already enabled in step 2.

Reply:
It is walker and we return to Hash join

Reply:
And the main question here, will we break something that previously depended on prefetch_inner value, if we change it at apply_shareinput_dag_to_tree stage.

Reply:
so you suggest adding a flag to the ShareInputContext that we will enable when we meet Shared Scan and when we return from recursion, check whether the current node is HashJoin and if so, enable prefetch_inner for it?

Reply:
I don't see much point in saving the prefetch state before apply_shareinput_dag_to_tree since if we have a Shared Scan we must enable prefetch

Reply:
Yep, kinda. We can do it without save

Reply:
okay let's try this approach

Reply:
And check this question. 

> And the main question here, will we break something that previously depended on prefetch_inner value, if we change it at apply_shareinput_dag_to_tree stage.

Reply:
the function you changed is called only during apply_shareinput_dag_to_tree?

Reply:
at the planning stage, the `prefetch_inner` value is used only for the Nested loop in `shareinput_walker`, then in `motion_sanity_walker`, which is executed after `apply_shareinput_dag_to_tree`.
the function in which `motion_sanity_walker` is called:
https://github.com/arenadata/gpdb/blob/215d8d8f9fec39300e4eb84977ade4fce040ebc1/src/backend/optimizer/plan/planner.c#L401
usage for Nested Loop in `shareinput_walker`:
https://github.com/arenadata/gpdb/blob/215d8d8f9fec39300e4eb84977ade4fce040ebc1/src/backend/cdb/cdbmutate.c#L1899
therefore, changing the `prefetch` for HashJoin in `apply_shareinput_dag_to_tree` looks safe.

Reply:
ORCA tests have failed, why?
Is there can be a case when HashJoin does not require the prefetch (there is no Shared Scans under that), but you enable prefetch based on condition `ctxt->producer_count > 0`, which is not related to that HashJoin (it could be set earlier before the walker got to the HashJoin.

Reply:
> ORCA tests have failed, why?

Known flaky dispatch test was failed.

Reply:
if we want to enable prefetch_inner only for those hashjoins in the plan that have a Shred Scan under them, then we can do this. add a flag to ApplyShareInputContext saying that there is a SharedScan below. turn it off when descending and turn it on when ascending if we meet SharedScan. And enable prefetch_inner based on this flag.

Reply:
>if we want to enable prefetch_inner only for those hashjoins in the plan that have a Shred Scan under them

Do we want that? 
Can the case I described  be possible (when we have bottleneck join without shared scans, but somewhere in the plan we have shared scans)?

>add a flag to ApplyShareInputContext

Or just check the difference between producer_count before the HashJoin and after it.

Reply:
> Or just check the difference between producer_count before the HashJoin and after it.

in this case, we need to store somewhere the number of producers at the moment when we met it.

Reply:
Whats does that condition mean?

Reply:
if there is no motion and/or CTE below HashJoin, then prefetch can also be turned off

Reply:
I don't get it, explain the motivation behing this condition

Reply:
prefetch_inner for Joins is enabled if the motionHazard flag is set for the underlying path, it is set in the case of Motion or CTE. therefore, if there is neither one nor the other, then it will be turned off. we have one situation when it additionally turns off in the case of BottleNeck. without this condition, we will enable it for all hashjoins even if it is not required.

Reply:
> without this condition, we will enable it for all hashjoins even if it is not required

Well, if you really want to follow this logic, you must check the case when HashJoin is at Bottleneck, but prefetch is not required.

Reply:
Ok, then write proper patch description.
And write proper code comments, because `even if` does not tell anything for people out of context.
```
/*
 * HashJoin requires prefetch_inner to be enabled if there is a
 * SharedScan in the plan even if it is on bottleneck.
 */
```

Reply:
>Can there be a case when HashJoin does not require the prefetch (there is no Shared Scans under that), but you enable prefetch based on condition ctxt->producer_count > 0, which is not related to that HashJoin (it could be set earlier before the walker got to the HashJoin)

And why not to do that, if your main idea is to make the changes cover as less cases as possible. Let's restrict the prefetch overwrite and do i tonly for bottleneck joins, that have Shared Scans **under** them.

Reply:
May you wrtie the Bottleneck world in a uniform manner in PR description?

Reply:
in the source code, it is written 'Bottleneck' so I think I'll correct the description.

Reply:
if there are Shared Scans in
the plan and Join **are** on bottleneck.

Fix that

Reply:
when **travelling** the plan in shareinput_mutator_dag_to_tree if there are Shared Scans in
the plan and the HashJoin is on bottleneck.

Reply:
Tests